unit-tests/sw-dev/test-units.py

```diff
@@ -0,0 +1,55 @@
+# License: Apache 2.0. See LICENSE file in root directory.
+# Copyright(c) 2023 Intel Corporation. All Rights Reserved.
+
+import pyrealsense2 as rs
+from rspy import log, test
+import sw
+
+    
+with sw.sensor( "Stereo Module" ) as sensor:
+    depth = sensor.video_stream( "Depth", rs.stream.depth, rs.format.z16 )
+    sensor.start( depth )
+
+    with test.closure( "By default, frames do not have units" ):
+        f = depth.frame()
+        test.check_equal( f.depth_units, 0. )
+
+    # Publish it
+    f = sensor.publish( f )
+
+    with test.closure( "rs.stream.depth should generate depth-frames", on_fail=test.ABORT ):
+        df = rs.depth_frame( f )
+        test.check( df )
+
+    with test.closure( "No DEPTH_UNITS; Units should be 0" ):
+        test.check_false( sensor.supports( rs.option.depth_units ) )
+        test.check_equal( df.get_units(), 0. )
+        test.check_equal( df.get_distance( 0, 0 ), 0. )
+
+    with test.closure( "Set the sensor DEPTH_UNITS" ):
+        sensor.add_option( rs.option.depth_units, rs.option_range( 0, 1, 0.000001, 0.001 ), True )
+        #test.check_equal( sensor.get_option( rs.option.depth_units ), 0.001 )   it's not set yet
+        sensor.set_option( rs.option.depth_units, 0.001 )
+
+    with test.closure( "Frame units should not change after the fact" ):
+        sensor.set_option( rs.option.depth_units, 0.001 )
+        test.check_equal( df.get_units(), 0. )
+        test.check_equal( df.get_distance( 0, 0 ), 0. )
+
+    with test.closure( "But if we generate a new frame..." ):
+        f = depth.frame()
+        test.check_equal( f.depth_units, 0. )
+
+    # Publish it
+    f = sensor.publish( f )
+    df = rs.depth_frame( f )
+
+    with test.closure( "New frame should pick up DEPTH_UNITS" ):
+        test.check_approx_abs( df.get_units(), 0.001, 0.00000001 )
+        test.check_approx_abs( df.get_distance( 0, 0 ), 26.985, 0.000001 )
```

Comment:
So the distance did change when setting the depth_units?
Where did you modified it from 0?

Reply:
From the sensor: we change the sensor `DEPTH_UNITS`, to the new frame picked it up.

Reply:
I will ask again here as my question before was general.
Do we use the depth units in the frame calculation? for some scale factor?
How did you get the 26.985? there is no real sensor here, so this value comes from somewhere and I am missing it..

Reply:
Yes: see [here](https://github.com/IntelRealSense/librealsense/blob/development/src/frame.cpp#L258).
We do `pixel-value * depth-units`, in this case 0x6969 (from sw.py buffer) * 0.001 = 26.985. That's why I changed sw.py to 0x69 -- it was 0x00 before, so would have yielded 0 distance.

Reply:
Now I understand thanks.
Can you add a comment before this check? its not clear as the initialization is done in another file.

What will happen if no depth unit is set? depth will always be zero?

I am not sure its the best default behavior as the user doesnt know he must set it.
What do you think?

Reply:
Without depth units, there can be no depth. With regular cameras, we take care to set this (it wasn't always a setting), but with software sensors it has to be provided by someone...

Before, it was completely uninitialized with no way to set it at all... I'd say this is an improvement. And again, this is only for the `get_distance()` function -- there are other ways to get at the pixel data and calculating depth. I just don't think SW devices saw much use.

Reply:
(it was uninitialized in the wrappers; in C++ it's part of the structure so visible to all and settable)

Reply:
I agree its good to fix in the root.
Only issue is that when user will use without setting the depth unit he we get 0.

Maybe we need to set it to 1? no impact on the pixels by default?

Reply:
See my comment offline. I do not believe we can or should tell the user what value to put there. It is up to the user to correctly set the data.